src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+using SafeCurlHandle = Interop.libcurl.SafeCurlHandle;
+using CURLoption = Interop.libcurl.CURLoption;
+using CURLcode = Interop.libcurl.CURLcode;
+using CURLINFO = Interop.libcurl.CURLINFO;
+
+namespace System.Net.Http
+{
+    internal class CurlHandler : HttpMessageHandler
+    {
+        #region Constants
+        private const string UriSchemeHttps = "https";
+        #endregion
+
+        #region Fields
+        private volatile bool _anyOperationStarted;
+        private volatile bool _disposed;
+        private bool _automaticRedirection = true;
+        #endregion
+
+        internal CurlHandler()
+        {
+        }
+
+        #region Properties
+        internal bool AutomaticRedirection
+        {
+            get
+            {
+                return _automaticRedirection;
+            }
+
+            set
+            {
+                CheckDisposedOrStarted();
+                _automaticRedirection = value;
+            }
+        }
+        #endregion
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+            }
+
+            base.Dispose(disposing);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(
+            HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request", SR.net_http_handler_norequest);
+            }
+            if (request.RequestUri.Scheme == UriSchemeHttps)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTPS stack is not yet implemented");
+            }
+            if (request.Content != null)
+            {
+                throw NotImplemented.ByDesignWithMessage("HTTP requests with a body are not yet supported");
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+
+            try
+            {
+                Task.Factory.StartNew(
+                    StartRequest,
```

Comment:
This one's a bit tricky, because the method you want to call is an instance method.  Normally you'd handle that by passing ```this``` as the ```object state```, but you're already passing in ```state```, so the slots taken.  In such cases, you typically have three options:

1. Live with allocating the delegate.
2. Allocate a new object to store all of the state, e.g. a ```Tuple``` that wraps both ```this``` and ```state```.
3. Modify the existing state object to carry the additional state.

In this case, I think (3) is the best approach.  You'd change RequestState to have an additional property/field:
```C#
public CurlHandler Handler { get; set; }
```
Then above when setting up the state you'd store ```this``` into it:
```C#
state.Handler = this;
```
and then you'd change your StartNew call to look like this:
```C#
Task.Factory.StartNew(s => {
    RequestState rs = (RequestState)s;
    rs.Handler.StartRequest(rs);
}, state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
```
Since you're then passing in the correctly typed instance to StartRequest, you can modify StartRequest to accept a ```RequestState``` rather than ```object``` and eliminate the cast at the beginning of that method.